### ISSUE TITLE
Coq is incompatible with OCaml < 4.07 + a recent GCC

### ISSUE THREAD

**Issue by ppedrot** [coq/coq#12940]
After updating my Debian installation, I had the surprise to see that Coq was not compiling anymore, due to errors while linking C code.
```
cd kernel/byterun/ && \
"/home/pm/.opam/4.06.1+fp/bin/ocamlfind" ocamlmklib -oc coqrun coq_fix_code.o coq_memory.o coq_values.o coq_interp.o
/usr/bin/ld : coq_interp.o:/home/pm/.opam/4.06.1+fp/lib/ocaml/caml/major_gc.h:69 : définitions multiples de « caml_major_work_credit »; coq_memory.o:/home/pm/.opam/4.06.1+fp/lib/ocaml/caml/major_gc.h:69 : défini pour la première fois ici
/usr/bin/ld : coq_interp.o:/home/pm/.opam/4.06.1+fp/lib/ocaml/caml/major_gc.h:68 : définitions multiples de « caml_major_ring_index »; coq_memory.o:/home/pm/.opam/4.06.1+fp/lib/ocaml/caml/major_gc.h:68 : défini pour la première fois ici
/usr/bin/ld : coq_interp.o:/home/pm/.opam/4.06.1+fp/lib/ocaml/caml/major_gc.h:67 : définitions multiples de « caml_major_ring »; coq_memory.o:/home/pm/.opam/4.06.1+fp/lib/ocaml/caml/major_gc.h:67 : défini pour la première fois ici
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile.build:355 : kernel/byterun/libcoqrun.a] Erreur 2
```

The problem is actually described in https://github.com/ocaml/ocaml/pull/1770 and has been solved from 4.07 onwards. This issue was seemingly only affecting Clang then, but now it's also affecting GCC. Here is my configuration:

```
$ ocaml --version
The OCaml toplevel, version 4.06.1
$ gcc --version
gcc (Debian 10.2.0-5) 10.2.0
$ ld --version
GNU ld (GNU Binutils for Debian) 2.35
```

This can be solved by passing flags to GCC so that it doesn't produce this warnings, but I don't know how portable that is.

**Comment by ppedrot** [comment]
The obvious alternative is to bump the minimal required version of OCaml to 4.07, but beware of regressions such as described in https://github.com/ocaml/ocaml/issues/7826.

**Comment by Blaisorblade** [comment]
Oh, is this the GCC 10 -fcommon issue? OCaml 4.07 itself broke on GCC 10, and needed a patch (in the opam packages) to pass -fcommon to the C compiler.

**Comment by JasonGross** [comment]
> The obvious alternative is to bump the minimal required version of OCaml to 4.07, but beware of regressions such as described in [ocaml/ocaml#7826](https://github.com/ocaml/ocaml/issues/7826).

This would break fiat-crypto, I think, do to that performance regression (unless it's since been fixed in a newer version of OCaml?)

**Comment by ppedrot** [comment]
Fixed by #15947.
